plutus-use-cases/src/Plutus/Contracts/Uniswap.hs

```diff
@@ -0,0 +1,988 @@
+{-# LANGUAGE TemplateHaskell            #-}
+{-# LANGUAGE TypeApplications           #-}
+{-# LANGUAGE TypeFamilies               #-}
+{-# LANGUAGE TypeOperators              #-}
+
+-- | A decentralized exchange for arbitrary token pairs following the
+-- [Uniswap protocol](https://uniswap.org/whitepaper.pdf).
+--
+module Plutus.Contracts.Uniswap
+    ( Coin (..)
+    , coin, coinValueOf
+    , Uniswap (..), uniswap
+    , poolStateCoinFromUniswapCurrency, liquidityCoin
+    , CreateParams (..)
+    , SwapParams (..)
+    , CloseParams (..)
+    , RemoveParams (..)
+    , AddParams (..)
+    , UniswapUserSchema, UserContractState (..)
+    , UniswapOwnerSchema
+    , start, create, add, remove, close, swap, pools
+    , ownerEndpoint, userEndpoints
+    ) where
+
+import           Control.Monad                    hiding (fmap)
+import qualified Data.Map                         as Map
+import           Data.Monoid                      (Last (..))
+import           Data.Proxy                       (Proxy (..))
+import           Data.Text                        (Text, pack)
+import           Data.Void                        (Void)
+import           Ledger                           hiding (singleton)
+import           Ledger.Constraints               as Constraints
+import           Ledger.Constraints.OnChain       as Constraints
+import           Ledger.Constraints.TxConstraints as Constraints
+import qualified Ledger.Typed.Scripts             as Scripts
+import           Ledger.Value                     as Value
+import           Playground.Contract
+import           Plutus.Contract                  hiding (when)
+import qualified Plutus.Contracts.Currency        as Currency
+import qualified PlutusTx
+import           PlutusTx.Prelude                 hiding (Semigroup (..), unless)
+import           Prelude                          (Semigroup (..))
+import qualified Prelude
+import           Text.Printf                      (printf)
+
+feeNum, feeDen :: Integer
+feeNum = 3
+feeDen = 1000
+
+uniswapTokenName, poolStateTokenName :: TokenName
+uniswapTokenName = "Uniswap"
+poolStateTokenName = "Pool State"
+
+-- | A pair consisting of a 'CurrencySymbol' and a 'TokenName'.
+-- Coins are the entities that can be swapped in the exchange.
+data Coin = Coin
+    { cCurrency :: CurrencySymbol
+    , cToken    :: TokenName
+    } deriving (Show, Generic, ToJSON, FromJSON, ToSchema)
+
+PlutusTx.unstableMakeIsData ''Coin
+PlutusTx.makeLift ''Coin
+
+instance Eq Coin where
+    {-# INLINABLE (==) #-}
+    c == d = cCurrency c == cCurrency d && cToken c == cToken d
+
+instance Prelude.Eq Coin where
+    (==) = (==)
+
+{-# INLINABLE compareCoins #-}
+compareCoins :: Coin -> Coin -> Ordering
+compareCoins c d = case compare (cCurrency c) (cCurrency d) of
+    LT -> LT
+    GT -> GT
+    EQ -> compare (cToken c) (cToken d)
+
+instance Prelude.Ord Coin where
+    compare = compareCoins
+
+{-# INLINABLE coinLT #-}
+coinLT :: Coin -> Coin -> Bool
```

Comment:
Really wish we had that `PartialOrd` class already...

Reply:
I just replaced this with `<` ; I think it works fine

Reply:
Hmm, although that will say that `(Bar, 1) < (Foo, 2)`, whereas `coinLT` won't. i.e. if you feed it incompatible tokens then `coinLT` is stricter. _Maybe_ this doesn't matter, but we'd have to be more careful, perhaps?